### PR TITLE
Fix double free when using stream bitmaps with hash bitmaps.

### PR DESCRIPTION
**Pull request by xinzweb**
Fixes double free issue when we use two different types (i.e. Stream and Hash) of bitmap indexes under an AND/OR Bitmap operator.

Consider a query like `select count(1) from bmcrash where bitmap_col = '999' AND btree_col1 = 'abcdefg999';` 
where we have a Btree index on `btree_col1` and a `BitmapIndex` on `bitmap_col`.

```
                                              QUERY PLAN
------------------------------------------------------------------------------------------------------
 Aggregate  (cost=0.00..0.01 rows=1 width=8)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..0.01 rows=1 width=8)
         ->  Aggregate  (cost=0.00..0.01 rows=1 width=8)
               ->  Bitmap Table Scan on bmcrash  (cost=0.00..0.01 rows=1 width=1)
                     Recheck Cond: bitmap_col = '999'::text AND btree_col1::text = 'abcdefg999'::text
                     ->  BitmapAnd  (cost=0.00..0.00 rows=0 width=0)
                           ->  Bitmap Index Scan on bm_idx  (cost=0.00..0.00 rows=0 width=0)
                                 Index Cond: bitmap_col = '999'::text
                           ->  Bitmap Index Scan on bt_idx  (cost=0.00..0.00 rows=0 width=0)
                                 Index Cond: btree_col1::text = 'abcdefg999'::text
 Settings:  optimizer=on
 Optimizer status: PQO version 1.665
```

In the plan above, we have two `BitmapIndexScan` and a `BitmapAnd` operator that merges the two bitmaps, viz. a `StreamBitmap` (`bm_idx` on `bitmap_col`) 
and a `HashBitmap` (`bt_idx` on `btree_col1`) created at runtime from Btree.

When we have mixed type of bitmaps connected using AND or OR, we wrap the `HashBitmap` into a `StreamBitmap` and combine them in a `OpStreamBitmap` (e.g. `BitmapAnd`), 
which contains pointers to the original two bitmaps. With this scheme there are two pointers to the same `HashBitmap` 
(one from the `OpStreamBitmap` and another in the `BitmapIndexScan` of `bt_idx`).

So when it comes time to free the `HashBitmap`, it gets freed twice - once using `tbm_stream_free` directly from `ExecEndBitmapIndexScan`, 
and then once again via `opstream_free` when we free the `OpStreamBitmap`

ExecEndBitmapIndexScan -> tbm_bitmap_free -> tbm_free
ExecEndBitmapIndexScan -> tbm_bitmap_free -> **opstream_free** -> tbm_stream_free -> tbm_free

To recap, we try to free only once per executor operator. But, the `tidbitmap.c` API `tbm_bitmap_free` will call `tbm_free` on the `HashBitmap` first 
and subsequently for second `OpStreamBitmap` (as produced by appending `HashBitmap` to the StreamBitmap) will iterate through all its contained bitmaps 
and call `tbm_free` on the `HashBitmap` again. This will result in a double free.

To fix this we keep track of the ownership of the underlying `HashBitmap` which is kept in `HashBitmapOpaque`, and only free the `HashBitmap` if the caller owns it.

### CODE REVIEW

**Comment by xinzweb** [comment]
@foyzur please take a look and +1. Thanks.

**Comment by armenatzoglou** [comment]
Great job @xinzweb and @hardikar! This is a safe solution and follows the approach that postgres solves similar issues. 

+1

File: src/backend/nodes/tidbitmap.c [hunk 1]
```diff
@@ -1193,13 +1194,27 @@ tbm_create_stream_node(HashBitmap *tbm)
...
+	op->tofree = true;
 	op->entry = NULL;
 
 	is->opaque = (void *)op;
 
 	return is;
 }
 
+StreamNode *
+tbm_create_stream_node_ref(HashBitmap *tbm)
```
Comment (karthijrk) [inline src/backend/nodes/tidbitmap.c]: suffix _ref doesn't tell me that this is not taking the ownership. Can we have some other better name?

File: src/backend/nodes/tidbitmap.c [hunk 2]
```diff
@@ -1193,13 +1194,27 @@ tbm_create_stream_node(HashBitmap *tbm)
...
+{
+	IndexStream* sn = tbm_create_stream_node(tbm);
+	/*
+	 * Do not take ownership of the HashBitmap.
+	 */
+	HashStreamOpaque *op = (HashStreamOpaque*) sn->opaque;
+	op->tofree = false;
```
Comment (karthijrk) [inline src/backend/nodes/tidbitmap.c]: I am afraid of this boolean. What if the owner deletes the op but the other person who doesn't own and try to access it.

File: src/backend/nodes/tidbitmap.c (hunk 1 above)
Comment (xinzweb) [inline src/backend/nodes/tidbitmap.c]: What's your suggestion on new names?

File: src/backend/nodes/tidbitmap.c (hunk 2 above)
Comment (xinzweb) [inline src/backend/nodes/tidbitmap.c]: If that's the case, don't we get SEGV? Since, all ICG passed and no SEGV, until you got a repro, I will say our ICG said it's a good fix. :)

Comment (ashwinstar) [inline src/backend/nodes/tidbitmap.c]: If the ordering is 100% guaranteed then approach works. Ideally good to de-couple implementation from ordering assumption, so can a refcount be maintained  to track the references and once drops to zero free the same ?

Comment (karthijrk) [inline src/backend/nodes/tidbitmap.c]: I agree, that ICG pass and based on that there may be no issue today. But are we sure by design whoever set the flag as true are the real owner? Also by design are we guaranteed that no one else will de-refer the pointer once the owner deletes it ? 

I don't have full knowledge of the code but just want to make sure there is no potential bug because of this boolean. For e.g. a potential bug could be future code change(if we are not guaranteed by design) could result in deferring the pointer after the owner deletes it.

Comment (xinzweb) [inline src/backend/nodes/tidbitmap.c]: I am not sure I get the 'ordering' assumptions. It's not about who release first, and who release second. It's more about who actually own it. The owner should release it, but the references should not. 

In principle, I agree we should have refcount for shared objects, so that we can handle pointers better. The scope of this fix is to only focus on the release of Bitmap Index Scan.

In this scenario, the HashBitmap is only shared if we wrap it around using a StreamBitmap. In this case, all the StreamBitmap are references, and they should NOT free the bitmap at all. Only the HashBitmap, the original owner, should free it.

I hope this make the scenario clearer. Thanks.

Comment (karthijrk) [inline src/backend/nodes/tidbitmap.c]: @foyzur What do you thin ?

Comment (xinzweb) [inline src/backend/nodes/tidbitmap.c]: @karthijrk I agree with you in principle, and there is a but.

I don't have full context either, so the fix is isolated enough to only the scenarios we know we failed, e.g. AND and OR of Bitmap Index Scans. Hence, the delta is only in those two places, and all the existing logics are the same as before.

Since I don't have the FULL context, hence, I don't have a 'proof' to guarantee there is no further bugs introduced by this fix. As a separate story, to improve our code confidence, the code coverage should be 80%+ and tests should be reliable and focused. I believe this is the process beyond fix of this bug.

Thoughts? Thanks.

File: src/backend/nodes/tidbitmap.c [hunk 3]
```diff
@@ -96,6 +96,7 @@ struct HashBitmap
 typedef struct HashStreamOpaque 
 {
 	HashBitmap *tbm;
+	bool tofree;
```
Comment (asimrp) [inline src/backend/nodes/tidbitmap.c]: Can we add a comment briefly describing why this is needed?

**Comment by armenatzoglou** [comment]
I had the similar concerns regarding the ordering in the previous PR https://github.com/greenplum-db/gpdb/pull/1088. I gave a +1 to this approach, but I believe that @ashwinstar's approach is safer.

File: src/backend/nodes/tidbitmap.c (hunk 2 above)
Comment (asimrp) [inline src/backend/nodes/tidbitmap.c]: I'm not able to understand the concern here.  Isn't the call to `tbm_create_stream_node()`, right before dereferencing `op`, allocating `op` using `palloc()`?

File: src/backend/nodes/tidbitmap.c (hunk 3 above)
Comment (xinzweb) [inline src/backend/nodes/tidbitmap.c]: Good point. Let me add something like `flag to indicate whether this opaque pointer owns the HashBitmap tbm`.

File: src/test/regress/sql/bitmapscan_ao.sql [hunk 4]
```diff
@@ -151,6 +151,28 @@ and c1.language = 'ZHS') bar,
...
+CREATE INDEX bt_idx_2 ON bmcrash USING btree (btree_col2);
+
+INSERT INTO bmcrash (btree_col2, bitmap_col, btree_col1)
+SELECT date '2015-01-01' + (i % (365 * 2)), i % 1000, 'abcdefg' || (i% 1000)
+from generate_series(1,10000) as i;
+
+select count(1) from bmcrash where bitmap_col = '999' AND btree_col1 = 'abcdefg999';
+select count(1) from bmcrash where bitmap_col = '999' OR btree_col1 = 'abcdefg999';
+select count(1) from bmcrash where bitmap_col = '999' OR btree_col1 = 'abcdefg999' AND btree_col2 = '2015-01-01';
+select count(1) from bmcrash where bitmap_col = '999' AND btree_col1 = 'abcdefg999' OR btree_col2 = '2015-01-01';
```
Comment (asimrp) [inline src/test/regress/sql/bitmapscan_ao.sql]: How about adding a test with two `StreamBitmap`s connected with `AND` / `OR`, only to ensure that we are not breaking anything there?  Feel free to ignore this comment, I'm not sure if it makes sense.

File: src/backend/nodes/tidbitmap.c (hunk 2 above)
Comment (xinzweb) [inline src/backend/nodes/tidbitmap.c]: Yes, but we are not talking about free the `op`. We are talking about free up the `tbm` inside the `op`.

Comment (hardikar) [inline src/backend/nodes/tidbitmap.c]: I also wanted to mention that `ExecEndBitmapAnd` does not do any clean up, but rather iterates it's bitmap plans and calls `ExecEndNode(bitmapplans[i]);` The appropriate `ExecEndBitmapIndexScan` does the free-ing. Now this design tells me that it's the IndexScanNodes (of the Hashtable) that actually owns the underlying tbm object. And the other IndexScanNode (StreamBitmap) points to a Stream node that holds a "non-owning" reference to the tbm. 

As you can see, the ordering doesn't matter then - because the design "seems" to be to have the owners do the actual free-ing. Having a "_ref " version of the function, to me, makes it more explicit that it is a reference (not a owner).

File: src/test/regress/sql/bitmapscan_ao.sql (hunk 4 above)
Comment (xinzweb) [inline src/test/regress/sql/bitmapscan_ao.sql]: Can we also have operator count to verify that we have bitmap table scan and index scan?

Comment (hardikar) [inline src/test/regress/sql/bitmapscan_ao.sql]: +1. Having those doesn't hurt, we should check to see if those aren't already tested elsewhere.

File: src/backend/nodes/tidbitmap.c (hunk 2 above)
Comment (asimrp) [inline src/backend/nodes/tidbitmap.c]: Got it now, thanks.  Yes it is a valid concern.  Thinking of refcount option, it will suffer the same fate.  E.g. if we replace `bool tofree` with an `int refcount`, we have no way to guard against someone directly calling `tbm_free()` without reducing refcount.

File: src/backend/nodes/tidbitmap.c (hunk 3 above)
Comment (hardikar) [inline src/backend/nodes/tidbitmap.c]: @xinzweb I think we should change the name of this variable. `tofree` is too vague. This seems to confuse readers that it tracks whether op needs to be free'd, but the concern is more about the `tbm`. How about `tofreetbm` or `to_free_tbm`.?

**Comment by foyzur** [comment]
@karthijrk I think the ordering is fine. The way the plan gets generated is BitmapIndexScan will be at the very bottom, which is also supposed owner of the bitmap. And the way ExecutorEnd works, the current operator first cleans up and then calls the subtree cleanup (i.e., BitmapAnd and BitmapOr will have already done with their usage).

**Comment by asimrp** [comment]
@foyzur and @hardikar, the ordering is fine now.  I think the question is should the ordering change in future, will this PR be still valid?

**Comment by karthijrk** [comment]
In that case, I am fine with this. 👍

File: src/backend/nodes/tidbitmap.c (hunk 1 above)
Comment (karthijrk) [inline src/backend/nodes/tidbitmap.c]: I wish I can come up with some other better name. If you also don't have any alternative I am fine with this given that you added the comments.

**Comment by foyzur** [comment]
@asimrp I modified my comments. In addition to ExecEnd ordering, we still have safeguard that we are in clean up mode and therefore no one is accessing that bitmap and the composite bitmaps' intermediate nodes should be valid as those are not leaf nodes and owned by the valid owners. @hardikar and @xinzweb please share your thoughts on this too.

**Comment by foyzur** [comment]
LGTM.
